src/auxinfo/participant.rs

```diff
@@ -440,83 +452,91 @@ impl AuxInfoParticipant {
             .is_err()
         {
             self.stash_message(message)?;
             return Ok(ProcessOutcome::Incomplete);
         }
 
-        let global_rid: [u8; 32] =
-            self.storage
-                .retrieve(StorageType::GlobalRid, message.id(), self.id)?;
-        let decom: AuxInfoDecommit =
-            self.storage
-                .retrieve(StorageType::Decommit, message.id(), message.from())?;
+        let global_rid = self
+            .local_storage
+            .retrieve::<GlobalRidStorageType>(message.id(), self.id)?;
+        let decom = self
+            .local_storage
+            .retrieve::<DecommitStorageType>(message.id(), message.from())?;
 
-        let auxinfo_pub = decom.into_public();
+        let auxinfo_pub = decom.clone().into_public();
 
         let proof = AuxInfoProof::from_message(message)?;
         proof.verify(
             message.id(),
-            global_rid,
+            *global_rid,
             auxinfo_pub.params(),
             auxinfo_pub.pk().modulus(),
         )?;
 
-        self.storage.store(
-            StorageType::Public,
-            message.id(),
-            message.from(),
-            &auxinfo_pub,
-        )?;
+        self.local_storage
+            .store::<PublicStorageType>(message.id(), message.from(), auxinfo_pub);
 
         // Check if we've stored all the public auxinfo_pubs
-        let keyshare_done = self.storage.contains_for_all_ids(
-            StorageType::Public,
-            message.id(),
-            &[self.other_participant_ids.clone(), vec![self.id]].concat(),
-        )?;
+        let keyshare_done = self
+            .local_storage
+            .contains_for_all_ids::<PublicStorageType>(
+                message.id(),
+                &[self.other_participant_ids.clone(), vec![self.id]].concat(),
+            );
 
         // If so, we completed the protocol! Return the outputs.
         if keyshare_done {
             for oid in self.other_participant_ids.iter() {
-                self.storage.transfer::<StorageType, AuxInfoPublic>(
-                    main_storage,
-                    StorageType::Public,
+                let public = self
+                    .local_storage
+                    .retrieve::<PublicStorageType>(message.id(), *oid)?;
+                main_storage.store(
+                    PersistentStorageType::AuxInfoPublic,
                     message.id(),
                     *oid,
+                    public,
                 )?;
             }
-            self.storage.transfer::<StorageType, AuxInfoPublic>(
-                main_storage,
-                StorageType::Public,
+            let public = self
+                .local_storage
+                .retrieve::<PublicStorageType>(message.id(), self.id)?;
+            main_storage.store(
+                PersistentStorageType::AuxInfoPublic,
                 message.id(),
                 self.id,
+                public,
             )?;
```

Comment:
Oh, sorry! If you scroll down a few lines, when I added the returned outputs, I combined the `AuxInfoPublic` retrievals for `other_participant_ids` and `self.id` into one group with the `all_participants()` method. I got confused and thought I had also done it for the transfers, but clearly not. You could still combine them, like
```rust
for pid in self.all_participants() {
   let public = self.local_storage.retrieve(...);
   main_storage.store(...)
}
```
Not a huge deal since this is going away, but this other-self duplication pattern is all over and I'm on a minor crusade to remove it.

Reply:
Ah I see! Yes, that's much nicer.